# Module 1 homework

The goal of this homework is to train a simple model for predicting the duration of a ride - similar to what we did in this module.

## Q1. Downloading the data

We'll use [the same NYC taxi dataset](https://www1.nyc.gov/site/tlc/about/tlc-trip-record-data.page),
but instead of **Green** Taxi Trip Records, we'll use **Yellow** Taxi Trip Records.

Download the data for January and February 2022.

In [1]:
!wget -NP ./data https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2022-01.parquet
!wget -NP ./data https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2022-02.parquet

--2023-05-18 17:56:53--  https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2022-01.parquet
Resolviendo d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 2600:9000:21d6:8000:b:20a5:b140:21, 2600:9000:21d6:1000:b:20a5:b140:21, 2600:9000:21d6:3c00:b:20a5:b140:21, ...
Conectando con d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)[2600:9000:21d6:8000:b:20a5:b140:21]:443... conectado.
Petición HTTP enviada, esperando respuesta... 304 Not Modified
Fichero “./data/yellow_tripdata_2022-01.parquet” no modificado en el servidor. Omitiendo la descarga.

--2023-05-18 17:56:53--  https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2022-02.parquet
Resolviendo d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 2600:9000:21d6:d200:b:20a5:b140:21, 2600:9000:21d6:e200:b:20a5:b140:21, 2600:9000:21d6:e600:b:20a5:b140:21, ...
Conectando con d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)[2600:9000:21d6:d200:b:20a5:b140:21]:44

**Read the data for January. How many columns are there?**

In [2]:
import pandas as pd

In [3]:
df = pd.read_parquet("./data/yellow_tripdata_2022-01.parquet")
df

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count   
0               1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0  \
1               1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2               2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3               2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4               2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
...           ...                  ...                   ...              ...   
2463926         2  2022-01-31 23:36:53   2022-01-31 23:42:51              NaN   
2463927         2  2022-01-31 23:44:22   2022-01-31 23:55:01              NaN   
2463928         2  2022-01-31 23:39:00   2022-01-31 23:50:00              NaN   
2463929         2  2022-01-31 23:36:42   2022-01-31 23:48:45              NaN   
2463930         2  2022-01-31 23:46:00   2022-02-01 00:13:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID   
0                 3.80         1.0                  N           142  \
1                 2.10         1.0                  N           236   
2                 0.97         1.0                  N           166   
3                 1.09         1.0                  N           114   
4                 4.30         1.0                  N            68   
...                ...         ...                ...           ...   
2463926           1.32         NaN               None            90   
2463927           4.19         NaN               None           107   
2463928           2.10         NaN               None           113   
2463929           2.92         NaN               None           148   
2463930           8.94         NaN               None           186   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount   
0                 236             1        14.50    3.0      0.5        3.65  \
1                  42             1         8.00    0.5      0.5        4.00   
2                 166             1         7.50    0.5      0.5        1.76   
3                  68             2         8.00    0.5      0.5        0.00   
4                 163             1        23.50    0.5      0.5        3.00   
...               ...           ...          ...    ...      ...         ...   
2463926           170             0         8.00    0.0      0.5        2.39   
2463927            75             0        16.80    0.0      0.5        4.35   
2463928           246             0        11.22    0.0      0.5        2.00   
2463929           164             0        12.40    0.0      0.5        0.00   
2463930           181             0        25.48    0.0      0.5        6.28   

         tolls_amount  improvement_surcharge  total_amount   
0                 0.0                    0.3         21.95  \
1                 0.0                    0.3         13.30   
2                 0.0                    0.3         10.56   
3                 0.0                    0.3         11.80   
4                 0.0                    0.3         30.30   
...               ...                    ...           ...   
2463926           0.0                    0.3         13.69   
2463927           0.0                    0.3         24.45   
2463928           0.0                    0.3         16.52   
2463929           0.0                    0.3         15.70   
2463930           0.0                    0.3         35.06   

         congestion_surcharge  airport_fee  
0                         2.5          0.0  
1                         0.0          0.0  
2                         0.0          0.0  
3                         2.5          0.0  
4                         2.5          0.0  
...                       ...          ...  
2463926                   NaN          NaN  
2463927                   NaN          NaN  
2463928                   NaN          NaN  
2463929                   N

> **Answer:** 19 columns.

## Q2. Computing duration

Now let's compute the `duration` variable. It should contain the duration of a ride in minutes.

In [4]:
df["duration"] = df.tpep_dropoff_datetime - df.tpep_pickup_datetime
df.duration = df.duration.apply(lambda td: td.total_seconds() / 60)
df

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count   
0               1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0  \
1               1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2               2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3               2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4               2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
...           ...                  ...                   ...              ...   
2463926         2  2022-01-31 23:36:53   2022-01-31 23:42:51              NaN   
2463927         2  2022-01-31 23:44:22   2022-01-31 23:55:01              NaN   
2463928         2  2022-01-31 23:39:00   2022-01-31 23:50:00              NaN   
2463929         2  2022-01-31 23:36:42   2022-01-31 23:48:45              NaN   
2463930         2  2022-01-31 23:46:00   2022-02-01 00:13:00              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID   
0                 3.80         1.0                  N           142  \
1                 2.10         1.0                  N           236   
2                 0.97         1.0                  N           166   
3                 1.09         1.0                  N           114   
4                 4.30         1.0                  N            68   
...                ...         ...                ...           ...   
2463926           1.32         NaN               None            90   
2463927           4.19         NaN               None           107   
2463928           2.10         NaN               None           113   
2463929           2.92         NaN               None           148   
2463930           8.94         NaN               None           186   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount   
0                 236             1        14.50    3.0      0.5        3.65  \
1                  42             1         8.00    0.5      0.5        4.00   
2                 166             1         7.50    0.5      0.5        1.76   
3                  68             2         8.00    0.5      0.5        0.00   
4                 163             1        23.50    0.5      0.5        3.00   
...               ...           ...          ...    ...      ...         ...   
2463926           170             0         8.00    0.0      0.5        2.39   
2463927            75             0        16.80    0.0      0.5        4.35   
2463928           246             0        11.22    0.0      0.5        2.00   
2463929           164             0        12.40    0.0      0.5        0.00   
2463930           181             0        25.48    0.0      0.5        6.28   

         tolls_amount  improvement_surcharge  total_amount   
0                 0.0                    0.3         21.95  \
1                 0.0                    0.3         13.30   
2                 0.0                    0.3         10.56   
3                 0.0                    0.3         11.80   
4                 0.0                    0.3         30.30   
...               ...                    ...           ...   
2463926           0.0                    0.3         13.69   
2463927           0.0                    0.3         24.45   
2463928           0.0                    0.3         16.52   
2463929           0.0                    0.3         15.70   
2463930           0.0                    0.3         35.06   

         congestion_surcharge  airport_fee   duration  
0                         2.5          0.0  17.816667  
1                         0.0          0.0   8.400000  
2                         0.0          0.0   8.966667  
3                         2.5          0.0  10.033333  
4                         2.5          0.0  37.533333  
...                       ...          ...        ...  
2463926                   NaN          NaN   5.966667  
2463927                   NaN

**What's the standard deviation of the trips duration in January?**

In [5]:
df.duration.std()

46.44530513776802

> **Answer:** 46.45 minutes.

## Q3. Dropping outliers

Next, we need to check the distribution of the `duration` variable. There are some outliers. Let's remove them and keep only the records where the duration was between 1 and 60 minutes (inclusive).

In [6]:
n_records_before = df.shape[0]
n_records_before

2463931

In [7]:
df = df[(df.duration >= 1) & (df.duration <= 60)]
n_records_after = df.shape[0]
n_records_after

2421440

**What fraction of the records left after you dropped the outliers?**

In [8]:
n_records_after / n_records_before

0.9827547930522406

> **Answer:** 98%

## Q4. One-hot encoding

Let's apply one-hot encoding to the pickup and dropoff location IDs. We'll use only these two features for our model.

* Turn the dataframe into a list of dictionaries
* Fit a dictionary vectorizer 
* Get a feature matrix from it

In [9]:
from sklearn.feature_extraction import DictVectorizer

In [10]:
categorical = ["PULocationID", "DOLocationID"]
df[categorical].dtypes

PULocationID    int64
DOLocationID    int64
dtype: object

In [11]:
df.loc[:, categorical] = df.loc[:, categorical].astype(str)
df[categorical].dtypes

PULocationID    object
DOLocationID    object
dtype: object

In [12]:
train_dicts = df[categorical].to_dict(orient="records")
train_dicts[:5]

[{'PULocationID': '142', 'DOLocationID': '236'},
 {'PULocationID': '236', 'DOLocationID': '42'},
 {'PULocationID': '166', 'DOLocationID': '166'},
 {'PULocationID': '114', 'DOLocationID': '68'},
 {'PULocationID': '68', 'DOLocationID': '163'}]

In [13]:
dv = DictVectorizer()
X_train = dv.fit_transform(train_dicts)
X_train

<2421440x515 sparse matrix of type '<class 'numpy.float64'>'
	with 4842880 stored elements in Compressed Sparse Row format>

**What's the dimensionality of this matrix (number of columns)?**

In [14]:
X_train.get_shape()

(2421440, 515)

> **Answer:** 515 columns.

## Q5. Training a model

Now let's use the feature matrix from the previous step to train a model. 

* Train a plain linear regression model with default parameters.
* Calculate the RMSE of the model on the training data.

In [15]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [16]:
target = "duration"

y_train = df[target].values

lr = LinearRegression()
lr.fit(X_train, y_train)

y_pred = lr.predict(X_train)

**What's the RMSE on train?**

In [17]:
mean_squared_error(y_train, y_pred, squared=False)

6.98619123059128

> **Answer:** 6.99 minutes.

## Q6. Evaluating the model

Now let's apply this model to the validation dataset (February 2022).

In [18]:
# data preprocessing
df_val = pd.read_parquet("./data/yellow_tripdata_2022-02.parquet")
df_val["duration"] = df_val.tpep_dropoff_datetime - df_val.tpep_pickup_datetime
df_val.duration = df_val.duration.apply(lambda td: td.total_seconds() / 60)
df_val = df_val[(df_val.duration >= 1) & (df_val.duration <= 60)]
df_val.loc[:, categorical] = df_val.loc[:, categorical].astype(str)

In [19]:
val_dicts = df_val[categorical].to_dict(orient="records")

X_val = dv.transform(val_dicts)
y_val = df_val[target].values

y_pred = lr.predict(X_val)

**What's the RMSE on validation?**

In [20]:
mean_squared_error(y_val, y_pred, squared=False)

7.786413522518241

> **Answer:** 7.79 minutes.